### PR TITLE
[STORM-1642] Catch Exception when deserialization failed. 

### PR DESCRIPTION
**Pull request by liurenjie1024**
This patch aims to fix bug STORM-1642.  When we failed to deserialize a tuple, we should skip this tuple and print a log rather than just shutdown the work.

### CODE REVIEW

**Comment by abhishekagarwal87** [comment]
How was deserialization failure resulting in NPE? Other than that, deserialization failure is the problem with an application itself which application should correct by registering correct serializer/deserializer. storm framework should not skip such tuples on its own.

**Comment by liurenjie1024** [comment]
As I have mentioned in bug report, the reason why NPE happens is that storm is vulnerable to fraud message from  processes outside the cluster. To reproduce the NPE, you just need to send a message [taskid 0] to [host]:[port] from anywhere, where taskid is id of one of the tasks running on [host]:[port]. In this case, storm will return a TaskMessage with payload set to null to the deserializer. 
Storm does not check whether the task message is from processes within the cluster, so deserialization may fail. I  think storm should skip the fruad task message rather than shutdowning the worker when deserialization failed.

**Comment by abhishekagarwal87** [comment]
I can't think of a reason of why would such a message be sent to storm from outside the cluster. Ideally only cluster machines and daemons should have access to the worker ports. Or if it indeeds needs to be solved, then a better method would be to ignore the zero length payload and not add `new TaskMessage(task, null)`.  Before you make any change, let's wait for comments from someone who is more familiar with this part of code.

**Comment by liurenjie1024** [comment]
The reason why such a message is sent is that our security team is scanning all the ports listened on each server and tries to detect potential weakness.
At first thought I just wanted to ignore messages with zero payload, but I think ignoring messages failed to be deserialized would be more robust. I think this also matches storm's design since it throws away messages whose target task id is not in the worker and this patch is just a complement.

File: storm-core/src/clj/backtype/storm/daemon/executor.clj [hunk 1]
```diff
@@ -431,13 +431,14 @@
...
-              (fast-list-iter [task-id task-ids]
+            (if tuple
+              ((when debug? (log-message "Processing received message FOR " task-id " TUPLE: " tuple))
+              (if task-id
                 (tuple-action-fn task-id tuple)
-                ))
+                ;; null task ids are broadcast tuples
+                (fast-list-iter [task-id task-ids]
+                  (tuple-action-fn task-id tuple)
+                  ))))
```
Comment (knusbaum) [inline storm-core/src/clj/backtype/storm/daemon/executor.clj]: Lots of parenthetical and formatting problems here.

**Comment by knusbaum** [comment]
Why not no both? I don't see any reason to propagate `TaskMessage`s with null payloads, but I agree we should be dropping tuples that failed to deserialize in the iterator. 

This is very much on the critical path, so someone else might want to weigh in.

**Comment by liurenjie1024** [comment]
NPE is just a special case of deserialization failure, right?

**Comment by knusbaum** [comment]
What do you mean?

If you mean null payloads are a special case, I disagree. In fact, it's not a failure at all. We received a null message, so we can perform a NOP as soon as possible.

**Comment by liurenjie1024** [comment]
Yes, I agree with you that we should drop a task message with null payload as soon as possible, and I'll push another commit to drop task messages with null payload.

File: storm-core/src/jvm/backtype/storm/messaging/netty/MessageBatch.java [hunk 2]
```diff
@@ -112,6 +115,9 @@ private void writeTaskMessage(ChannelBufferOutputStream bout, TaskMessage messag
             throw new RuntimeException("Task ID should not exceed "+Short.MAX_VALUE);
         
         bout.writeShort((short)task_id);
+        if (payload_len == 0) {
+            LOG.warn("Zero length payload to task {}.", (short)task_id);
```
Comment (gzm55) [inline storm-core/src/jvm/backtype/storm/messaging/netty/MessageBatch.java]: is this a debugging log? should remove them?

**Comment by knusbaum** [comment]
If you want to clean up the log messages, this still looks good. We'd want this instead to go into master and 1.x, though.

**Pull request closed without merging**